src/app_service/service/collectible/service.nim

```diff
@@ -1,47 +1,77 @@
-import chronicles, sequtils, json
+import NimQml, chronicles, sequtils, json
+import ../../../app/core/eventemitter
+import ../../../app/core/tasks/[qt, threadpool]
 
 import dto
 import ../network/service as network_service
 
 import ../../../backend/backend
 
+include ../../common/json_utils
+include async_tasks
+
 export dto
 
 logScope:
   topics = "collectible-service"
 
 const limit = 200
 
+const GetCollections* = "get-collections"
+
 type
-  Service* = ref object of RootObj
-    networkService: network_service.Service
-
-proc delete*(self: Service) =
-  discard
-
-proc newService*(networkService: network_service.Service): Service =
-  result = Service()
-  result.networkService = networkService
-
-proc init*(self: Service) =
-  discard
-
-proc getCollections*(self: Service, address: string): seq[CollectionDto] =
-  try:
-    let chainId = self.networkService.getNetworkForCollectibles().chainId
-    let response = backend.getOpenseaCollectionsByOwner(chainId, address)
-    return map(response.result.getElems(), proc(x: JsonNode): CollectionDto = x.toCollectionDto())
-  except Exception as e:
-    let errDesription = e.msg
-    error "error: ", errDesription
-    return
-
-proc getCollectibles*(self: Service, address: string, collectionSlug: string): seq[CollectibleDto] =
-  try:
-    let chainId = self.networkService.getNetworkForCollectibles().chainId
-    let response = backend.getOpenseaAssetsByOwnerAndCollection(chainId, address, collectionSlug, limit)
-    return map(response.result.getElems(), proc(x: JsonNode): CollectibleDto = x.toCollectibleDto())
-  except Exception as e:
-    let errDesription = e.msg
-    error "error: ", errDesription
-    return
+  GetCollectionsArgs* = ref object of Args
+    collections*: seq[CollectionDto]
+
+QtObject:
+  type
+    Service* = ref object of QObject
+      events: EventEmitter
+      threadpool: ThreadPool
+      networkService: network_service.Service
+
+  proc delete*(self: Service) =
+      self.QObject.delete
+
+  proc newService*(
+    events: EventEmitter,
+    threadpool: ThreadPool,
+    networkService: network_service.Service,
+  ): Service =
+    result = Service()
+    result.QObject.setup
+    result.events = events
+    result.threadpool = threadpool
+    result.networkService = networkService
+
+  proc init*(self: Service) =
+    discard
+
+  proc onGetCollections*(self: Service, response: string) {.slot.} =
+    let responseObj = response.parseJson
```

Comment:
@alaibe this is an operation which may fail (in case of any incorrectly structured json) and crash the app, please add try/catch block for it.
Especially cause from the line `15` from `async_tasks.nim` you're returning an empty string which will throw an error in `parseJson` function for sure.

Reply:
@alaibe here is a short and quick example, showing that: https://play.nim-lang.org/#ix=4fWQ